### ISSUE TITLE
[clang++] Frontend crashes

### ISSUE THREAD

**Issue by lionkor** [llvm/llvm-project#58667]
```console
clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

Clang version: latest on Arch Linux (from official Arch repos)

I didn't give this a better title because I have zero idea why. I hope this bug report is helpful, though I couldn't reproduce it any other way, so this is my entire project (sorry).

Output: 
```console
/usr/bin/ccache /usr/bin/clang++ -DDOCTEST_CONFIG_DISABLE -DNOISE_STATIC -I/home/lion/src/tribal/deps/libnoise/src -I/home/lion/src/tribal/deps/fmt/include -I/home/lion/src/tribal/deps/doctest -I/home/lion/src/tribal/deps/SFML/include -Og -std=gnu++20 -MD -MT CMakeFiles/game.dir/src/main.cpp.o -MF CMakeFiles/game.dir/src/main.cpp.o.d -o CMakeFiles/game.dir/src/main.cpp.o -c /home/lion/src/tribal/src/main.cpp
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /usr/bin/clang++ -Og -std=gnu++20 -DDOCTEST_CONFIG_DISABLE -DNOISE_STATIC -I/home/lion/src/tribal/deps/libnoise/src -I/home/lion/src/tribal/deps/fmt/include -I/home/lion/src/tribal/deps/doctest -I/home/lion/src/tribal/deps/SFML/include -c -MD -MT CMakeFiles/game.dir/src/main.cpp.o -MF CMakeFiles/game.dir/src/main.cpp.o.d -fcolor-diagnostics -o CMakeFiles/game.dir/src/main.cpp.o /home/lion/src/tribal/src/main.cpp
1.	/home/lion/src/tribal/src/main.cpp:123:92: current parser token ';'
2.	/home/lion/src/tribal/src/main.cpp:94:1: parsing struct/union/class body 'World'
3.	/home/lion/src/tribal/src/main.cpp:113:13: parsing function body 'World::World'
4.	/home/lion/src/tribal/src/main.cpp:113:13: in compound statement ('{}')
5.	/home/lion/src/tribal/src/main.cpp:120:64: in compound statement ('{}')
6.	/home/lion/src/tribal/src/main.cpp:121:68: in compound statement ('{}')
7.	/home/lion/src/tribal/src/main.cpp:122:34: lambda expression parsing
8.	/home/lion/src/tribal/src/main.cpp:122:78: in compound statement ('{}')
  #0 0x00007f256e4d545a (/usr/lib/libLLVM-14.so+0xcd545a)
  #1 0x00007f256e4d2aef llvm::sys::RunSignalHandlers() (/usr/lib/libLLVM-14.so+0xcd2aef)
  #2 0x00007f256e3bfb59 (/usr/lib/libLLVM-14.so+0xbbfb59)
  #3 0x00007f256d251a00 (/usr/lib/libc.so.6+0x38a00)
  #4 0x00007f25754f8d7e (/usr/lib/libclang-cpp.so.14+0xaf8d7e)
  #5 0x00007f25754fa6d7 (/usr/lib/libclang-cpp.so.14+0xafa6d7)
  #6 0x00007f25754fa994 (/usr/lib/libclang-cpp.so.14+0xafa994)
  #7 0x00007f25754f5ca4 (/usr/lib/libclang-cpp.so.14+0xaf5ca4)
  #8 0x00007f25754f90c1 (/usr/lib/libclang-cpp.so.14+0xaf90c1)
  #9 0x00007f25754fa6d7 (/usr/lib/libclang-cpp.so.14+0xafa6d7)
 #10 0x00007f25754fa994 (/usr/lib/libclang-cpp.so.14+0xafa994)
 #11 0x00007f257552b754 (/usr/lib/libclang-cpp.so.14+0xb2b754)
 #12 0x00007f25754f5ca4 (/usr/lib/libclang-cpp.so.14+0xaf5ca4)
 #13 0x00007f25754f90c1 (/usr/lib/libclang-cpp.so.14+0xaf90c1)
 #14 0x00007f25754fa6d7 (/usr/lib/libclang-cpp.so.14+0xafa6d7)
 #15 0x00007f25754fa994 (/usr/lib/libclang-cpp.so.14+0xafa994)
 #16 0x00007f2576f47304 (/usr/lib/libclang-cpp.so.14+0x2547304)
 #17 0x00007f25754e2243 (/usr/lib/libclang-cpp.so.14+0xae2243)
 #18 0x00007f2576f47232 (/usr/lib/libclang-cpp.so.14+0x2547232)
 #19 0x00007f25754e15dc (/usr/lib/libclang-cpp.so.14+0xae15dc)
 #20 0x00007f2576f47304 (/usr/lib/libclang-cpp.so.14+0x2547304)
 #21 0x00007f25754e2375 (/usr/lib/libclang-cpp.so.14+0xae2375)
 #22 0x00007f2576f47232 (/usr/lib/libclang-cpp.so.14+0x2547232)
 #23 0x00007f25754e15dc (/usr/lib/libclang-cpp.so.14+0xae15dc)
 #24 0x00007f2576f47304 (/usr/lib/libclang-cpp.so.14+0x2547304)
 #25 0x00007f25754e15dc (/usr/lib/libclang-cpp.so.14+0xae15dc)
 #26 0x00007f2576f47304 (/usr/lib/libclang-cpp.so.14+0x2547304)
 #27 0x00007f25754e2375 (/usr/lib/libclang-cpp.so.14+0xae2375)
 #28 0x00007f2576f47232 (/usr/lib/libclang-cpp.so.14+0x2547232)
 #29 0x00007f25754e15dc (/usr/lib/libclang-cpp.so.14+0xae15dc)
 #30 0x00007f2576f47304 (/usr/lib/libclang-cpp.so.14+0x2547304)
 #31 0x00007f25754e2375 (/usr/lib/libclang-cpp.so.14+0xae2375)
 #32 0x00007f2576f47232 (/usr/lib/libclang-cpp.so.14+0x2547232)
 #33 0x00007f25754e15dc (/usr/lib/libclang-cpp.so.14+0xae15dc)
 #34 0x00007f25754f5ca4 (/usr/lib/libclang-cpp.so.14+0xaf5ca4)
 #35 0x00007f25754fce19 clang::Expr::EvaluateAsInitializer(clang::APValue&, clang::ASTContext const&, clang::VarDecl const*, llvm::SmallVectorImpl<std::pair<clang::SourceLocation, clang::PartialDiagnostic> >&, bool) const (/usr/lib/libclang-cpp.so.14+0xafce19)
 #36 0x00007f2575459a04 clang::VarDecl::evaluateValueImpl(llvm::SmallVectorImpl<std::pair<clang::SourceLocation, clang::PartialDiagnostic> >&, bool) const (/usr/lib/libclang-cpp.so.14+0xa59a04)
 #37 0x00007f2575459bd9 clang::VarDecl::evaluateValue() const (/usr/lib/libclang-cpp.so.14+0xa59bd9)
 #38 0x00007f2575a71c5a clang::Sema::FinalizeVarWithDestructor(clang::VarDecl*, clang::RecordType const*) (/usr/lib/libclang-cpp.so.14+0x1071c5a)
 #39 0x00007f25759f1715 clang::Sema::CheckCompleteVariableDeclaration(clang::VarDecl*) (/usr/lib/libclang-cpp.so.14+0xff1715)
 #40 0x00007f25759e770b clang::Sema::AddInitializerToDecl(clang::Decl*, clang::Expr*, bool) (/usr/lib/libclang-cpp.so.14+0xfe770b)
 #41 0x00007f25751ef70a clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) (/usr/lib/libclang-cpp.so.14+0x7ef70a)
 #42 0x00007f257520ef0f clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::SourceLocation*, clang::Parser::ForRangeInit*) (/usr/lib/libclang-cpp.so.14+0x80ef0f)
 #43 0x00007f25752137f9 clang::Parser::ParseSimpleDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributesWithRange&, bool, clang::Parser::ForRangeInit*, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x8137f9)
 #44 0x00007f2575213c88 clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributesWithRange&, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x813c88)
 #45 0x00007f25752831cb clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x8831cb)
 #46 0x00007f2575283c64 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x883c64)
 #47 0x00007f2575288a02 clang::Parser::ParseCompoundStatementBody(bool) (/usr/lib/libclang-cpp.so.14+0x888a02)
 #48 0x00007f25752261e2 clang::Parser::ParseLambdaExpressionAfterIntroducer(clang::LambdaIntroducer&) (/usr/lib/libclang-cpp.so.14+0x8261e2)
 #49 0x00007f257523b2a3 clang::Parser::ParseLambdaExpression() (/usr/lib/libclang-cpp.so.14+0x83b2a3)
 #50 0x00007f25751fc4f3 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) (/usr/lib/libclang-cpp.so.14+0x7fc4f3)
 #51 0x00007f2576ec9f5d (/usr/lib/libclang-cpp.so.14+0x24c9f5d)
 #52 0x00007f2576eca0d6 (/usr/lib/libclang-cpp.so.14+0x24ca0d6)
 #53 0x00007f25752398a6 clang::Parser::ParseExpressionList(llvm::SmallVectorImpl<clang::Expr*>&, llvm::SmallVectorImpl<clang::SourceLocation>&, llvm::function_ref<void ()>) (/usr/lib/libclang-cpp.so.14+0x8398a6)
 #54 0x00007f25751f944a clang::Parser::ParsePostfixExpressionSuffix(clang::ActionResult<clang::Expr*, true>) (/usr/lib/libclang-cpp.so.14+0x7f944a)
 #55 0x00007f25751faf31 clang::Parser::ParseCastExpression(clang::Parser::CastParseKind, bool, bool&, clang::Parser::TypeCastState, bool, bool*) (/usr/lib/libclang-cpp.so.14+0x7faf31)
 #56 0x00007f2576ec9f5d (/usr/lib/libclang-cpp.so.14+0x24c9f5d)
 #57 0x00007f2576eca0d6 (/usr/lib/libclang-cpp.so.14+0x24ca0d6)
 #58 0x00007f2576eca22d (/usr/lib/libclang-cpp.so.14+0x24ca22d)
 #59 0x00007f25752853f0 clang::Parser::ParseExprStatement(clang::Parser::ParsedStmtContext) (/usr/lib/libclang-cpp.so.14+0x8853f0)
 #60 0x00007f257528331c clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x88331c)
 #61 0x00007f2575283c64 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x883c64)
 #62 0x00007f2575288a02 clang::Parser::ParseCompoundStatementBody(bool) (/usr/lib/libclang-cpp.so.14+0x888a02)
 #63 0x00007f2575288ee2 clang::Parser::ParseCompoundStatement(bool, unsigned int) (/usr/lib/libclang-cpp.so.14+0x888ee2)
 #64 0x00007f25752835e4 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x8835e4)
 #65 0x00007f2575283c64 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x883c64)
 #66 0x00007f2575283efb clang::Parser::ParseStatement(clang::SourceLocation*, clang::Parser::ParsedStmtContext) (/usr/lib/libclang-cpp.so.14+0x883efb)
 #67 0x00007f2575287631 clang::Parser::ParseForStatement(clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x887631)
 #68 0x00007f2575283632 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x883632)
 #69 0x00007f2575283c64 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x883c64)
 #70 0x00007f2575288a02 clang::Parser::ParseCompoundStatementBody(bool) (/usr/lib/libclang-cpp.so.14+0x888a02)
 #71 0x00007f2575288ee2 clang::Parser::ParseCompoundStatement(bool, unsigned int) (/usr/lib/libclang-cpp.so.14+0x888ee2)
 #72 0x00007f25752835e4 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x8835e4)
 #73 0x00007f2575283c64 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x883c64)
 #74 0x00007f2575283efb clang::Parser::ParseStatement(clang::SourceLocation*, clang::Parser::ParsedStmtContext) (/usr/lib/libclang-cpp.so.14+0x883efb)
 #75 0x00007f2575287631 clang::Parser::ParseForStatement(clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x887631)
 #76 0x00007f2575283632 clang::Parser::ParseStatementOrDeclarationAfterAttributes(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x883632)
 #77 0x00007f2575283c64 clang::Parser::ParseStatementOrDeclaration(llvm::SmallVector<clang::Stmt*, 32u>&, clang::Parser::ParsedStmtContext, clang::SourceLocation*) (/usr/lib/libclang-cpp.so.14+0x883c64)
 #78 0x00007f2575288a02 clang::Parser::ParseCompoundStatementBody(bool) (/usr/lib/libclang-cpp.so.14+0x888a02)
 #79 0x00007f257528a4ba clang::Parser::ParseFunctionStatementBody(clang::Decl*, clang::Parser::ParseScope&) (/usr/lib/libclang-cpp.so.14+0x88a4ba)
 #80 0x00007f25751dd8ca clang::Parser::ParseLexedMethodDef(clang::Parser::LexedMethod&) (/usr/lib/libclang-cpp.so.14+0x7dd8ca)
 #81 0x00007f25751db2b5 clang::Parser::ParseLexedMethodDefs(clang::Parser::ParsingClass&) (/usr/lib/libclang-cpp.so.14+0x7db2b5)
 #82 0x00007f257521a241 clang::Parser::ParseCXXMemberSpecification(clang::SourceLocation, clang::SourceLocation, clang::ParsedAttributesWithRange&, unsigned int, clang::Decl*) (/usr/lib/libclang-cpp.so.14+0x81a241)
 #83 0x00007f257521bb14 clang::Parser::ParseClassSpecifier(clang::tok::TokenKind, clang::SourceLocation, clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, bool, clang::Parser::DeclSpecContext, clang::ParsedAttributesWithRange&) (/usr/lib/libclang-cpp.so.14+0x81bb14)
 #84 0x00007f25751ff657 clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*) (/usr/lib/libclang-cpp.so.14+0x7ff657)
 #85 0x00007f2575279b53 clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/usr/lib/libclang-cpp.so.14+0x879b53)
 #86 0x00007f257527a418 (/usr/lib/libclang-cpp.so.14+0x87a418)
 #87 0x00007f257528f66e clang::Parser::ParseExternalDeclaration(clang::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (/usr/lib/libclang-cpp.so.14+0x88f66e)
 #88 0x00007f257528f9f6 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, bool) (/usr/lib/libclang-cpp.so.14+0x88f9f6)
 #89 0x00007f25751d9f18 clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib/libclang-cpp.so.14+0x7d9f18)
 #90 0x00007f25767b3919 clang::FrontendAction::Execute() (/usr/lib/libclang-cpp.so.14+0x1db3919)
 #91 0x00007f2576752e07 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib/libclang-cpp.so.14+0x1d52e07)
 #92 0x00007f257680a7ca clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/usr/lib/libclang-cpp.so.14+0x1e0a7ca)
 #93 0x000055c9132ea6f1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/usr/bin/clang+++0x176f1)
 #94 0x000055c9132ec337 (/usr/bin/clang+++0x19337)
 #95 0x00007f25764aa499 (/usr/lib/libclang-cpp.so.14+0x1aaa499)
 #96 0x00007f256e3bfc27 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/usr/lib/libLLVM-14.so+0xbbfc27)
 #97 0x00007f25764affff (/usr/lib/libclang-cpp.so.14+0x1aaffff)
 #98 0x00007f25764c5763 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/usr/lib/libclang-cpp.so.14+0x1ac5763)
 #99 0x00007f25764c5b39 clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/usr/lib/libclang-cpp.so.14+0x1ac5b39)
#100 0x00007f25764cde72 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/usr/lib/libclang-cpp.so.14+0x1acde72)
#101 0x000055c9132e146a main (/usr/bin/clang+++0xe46a)
#102 0x00007f256d23c290 (/usr/lib/libc.so.6+0x23290)
#103 0x00007f256d23c34a __libc_start_main (/usr/lib/libc.so.6+0x2334a)
#104 0x000055c9132e2655 _start (/usr/bin/clang+++0xf655)
clang-14: error: clang frontend command failed with exit code 139 (use -v to see invocation)
clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
clang-14: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-14: note: diagnostic msg: /tmp/main-6a824d.cpp
clang-14: note: diagnostic msg: /tmp/main-6a824d.sh
clang-14: note: diagnostic msg: 

********************
ninja: build stopped: subcommand failed.
```

Attached are the mentioned diagnostic messages:
[main-6a824d.zip](https://github.com/llvm/llvm-project/files/9888812/main-6a824d.zip)

The code mentioned specifically is 
```c++
struct World {
    static inline std::time_t seed = std::time(nullptr);
    static constexpr double NOISE_SCALE = 10000.0;
    static constexpr size_t TILE_SIZE = 32;
    struct Chunk {
        static constexpr size_t size = 20;
        std::array<std::array<std::string, /* width, x */ size>, /* height, y */ size> tiles;
    };

    World() {
        using namespace boost;
        std::mutex mtx;
        asio::thread_pool pool { 40 };

        constexpr size_t super_chunk = 1;

        for (size_t sx = 0; sx < chunks.w / super_chunk; ++sx) {
            for (size_t sy = 0; sy < chunks.h / super_chunk; ++sy) {
                asio::post(pool, [this, &mtx, super_chunk, sx, sy]() -> void {
                    std::array<std::array<Chunk, super_chunk>, super_chunk> local_chunks {};
                    std::string from_to = fmt::format("from ({}, {}) to ({}, {})", sx * super_chunk, sx * super_chunk, sx * super_chunk + super_chunk, sy * super_chunk + super_chunk);
                    l::info("generating superchunk {}", from_to);
                    for (size_t y = 0; y < super_chunk; ++y) {
                        for (size_t x = 0; x < super_chunk; ++x) {
                            local_chunks[y][x] = generate(x + sx * super_chunk, y + sy * super_chunk);
                        }
                    }
                    l::info("finished generating superchunk {}", from_to);
                    std::unique_lock lock(mtx);
                    for (size_t y = 0; y < super_chunk; ++y) {
                        for (size_t x = 0; x < super_chunk; ++x) {
                            chunks[{ x + sx * super_chunk, y + sy * super_chunk }] = std::move(local_chunks[y][x]);
                        }
                    }
                    l::info("saved superchunk {}", from_to);
                });
            }
        }
        pool.join();
    }
```

**Comment by lionkor** [comment]
Again, I'm sorry this is such a mess - Maybe it helps or can be solved based on this.

**Comment by shafik** [comment]
Looks like it could be related to: https://github.com/llvm/llvm-project/issues/50708

**Comment by cor3ntin** [comment]
@lionkor Due to lack of information, I'm closing this. If you still experience this bug, feel free to reopen the issue.

**Comment by Endilll** [comment]
Linked archive apparently contains data from a different crash of ancient Clang 3.0. I reduced it nevertheless:
```cpp
struct Set {}

Set::Set() {
  0,[
}

)
```
It was fixed in Clang 3.1: https://godbolt.org/z/6E13TK69j

During reduction yet another crash has surfaced:
```cpp
class Set {
  Set() {
    transform([] { (element; })
```
It was fixed in Clang 3.1 as well: https://godbolt.org/z/YW7G79G4s
